receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal/cpu_utilization_calculator.go

```diff
@@ -0,0 +1,106 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal"
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/shirou/gopsutil/v3/cpu"
+	"go.opentelemetry.io/collector/model/pdata"
+)
+
+var ErrInvalidElapsed = errors.New("invalid elapsed seconds")
+var ErrDifferentCPUs = errors.New("cannot compare times from different cpus")
+var ErrTimeStatNotFound = errors.New("cannot find TimesStat for cpu")
+
+// CPUUtilization stores the utilization percents [0-1] for the different cpu states
+type CPUUtilization struct {
+	CPU     string
+	User    float64
+	System  float64
+	Idle    float64
+	Nice    float64
+	Iowait  float64
+	Irq     float64
+	Softirq float64
+	Steal   float64
+}
+
+// CPUUtilizationCalculator calculates the cpu utilization percents for the different cpu states
+// It requires 2 []cpu.TimesStat and spend time to be able to calculate the difference
+type CPUUtilizationCalculator struct {
+	previousTime     pdata.Timestamp
+	previousCPUTimes []cpu.TimesStat
+}
+
+// CalculateAndRecord calculates the cpu utilization for the different cpu states comparing previously
+// stored []cpu.TimesStat and time.Time and current []cpu.TimesStat and current time.Time
+// If no previous data is stored it will return empty slice of CPUUtilization and no error
+func (c *CPUUtilizationCalculator) CalculateAndRecord(now pdata.Timestamp, cpuTimes []cpu.TimesStat, recorder func(pdata.Timestamp, CPUUtilization)) error {
+	elapsedSeconds := now.AsTime().Sub(c.previousTime.AsTime()).Seconds()
+
+	if c.previousCPUTimes != nil {
+		for _, previousCPUTime := range c.previousCPUTimes {
+			currentCPUTime, err := cpuTimeForCPU(previousCPUTime.CPU, cpuTimes)
+			if err != nil {
+				return fmt.Errorf("getting time for cpu %s: %w", previousCPUTime.CPU, err)
+			}
+			utilization, err := cpuUtilization(previousCPUTime, currentCPUTime, elapsedSeconds)
+			if err != nil {
+				return fmt.Errorf("getting utilization for cpu %s: %w", previousCPUTime.CPU, err)
+			}
+			recorder(now, utilization)
+		}
+	}
+	c.previousCPUTimes = cpuTimes
+	c.previousTime = now
+
+	return nil
+}
+
+// cpuUtilization calculates the difference between 2 cpu.TimesStat using spent time between them
+// If no time was spent between TimesStat an error will be returned
+// If TimesStats do not belog to same CPU an error will be returned
+func cpuUtilization(timeStart cpu.TimesStat, timeEnd cpu.TimesStat, elapsedSeconds float64) (CPUUtilization, error) {
+	if elapsedSeconds <= 0 {
+		return CPUUtilization{}, fmt.Errorf("%f: %w", elapsedSeconds, ErrInvalidElapsed)
+	}
+	if timeStart.CPU != timeEnd.CPU {
+		return CPUUtilization{}, fmt.Errorf("%s <> %s: %w", timeStart.CPU, timeEnd.CPU, ErrDifferentCPUs)
+	}
```

Comment:
Why this check is needed given that we have https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7130/files#diff-d05de2fda70869723c63a5de25781cf2d25861b7c6bd306b2ad6f3df4d180075R57 ?

Reply:
You are right, removing it thanks